libdnf/base/goal.cpp

```diff
@@ -704,20 +729,60 @@ void Goal::Impl::add_upgrades_distrosync_to_goal(Action action, const std::strin
     // TODO(jmracek) Apply security filters
     // TODO(jmracek) q = q.available().union(installed_query.latest())
     // Required for a correct upgrade of installonly packages
-    solv_map_to_id_queue(tmp_queue, *query.p_impl);
     switch (action) {
         case Action::UPGRADE:
+            solv_map_to_id_queue(tmp_queue, *query.p_impl);
             rpm_goal.add_upgrade(tmp_queue, best, clean_requirements_on_remove);
             break;
         case Action::DISTRO_SYNC:
+            solv_map_to_id_queue(tmp_queue, *query.p_impl);
             rpm_goal.add_distro_sync(tmp_queue, strict, best, clean_requirements_on_remove);
             break;
+        case Action::DOWNGRADE: {
+            query.ifilter_available().ifilter_downgrades();
+            Pool * pool = sack.p_impl->get_pool();
+            std::vector<Solvable *> tmp_solvables;
+            for (auto pkg_id : *query.p_impl) {
+                tmp_solvables.push_back(rpm::solv::get_solvable(pool, pkg_id));
+            }
+            std::sort(tmp_solvables.begin(), tmp_solvables.end(), nevra_solvable_cmp_key);
+            std::map<Id, std::vector<Id>> name_arches;
+            for (auto installed_id : *relevant_installed_na.p_impl) {
+                Solvable * solvable = rpm::solv::get_solvable(pool, installed_id);
+                auto & arches = name_arches[solvable->name];
```

Comment:
What is the meaning of the code?
Isn't the `arches` always an empty vector? And so `unique` will forever remain true.

Reply:
Thanks. I forget to add arch to arches when the arch is unique. Fixed.

Reply:
The code prevents creation multiple times the same job for the same name and arch. Mostly valid for duplicates and installonly packages.